SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/OwnershipLockingPolicy.cpp

```diff
@@ -0,0 +1,297 @@
+	if (IsExplicitlyLocked(Actor) || IsLockedHierarchyRoot(Actor))
+	{
+		return true;
+	}
+
+	// Is the hierarchy root of this Actor explicitly locked or on a locked hierarchy ownership path.
+	if (AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor))
+	{
+		return IsExplicitlyLocked(HierarchyRoot) || IsLockedHierarchyRoot(HierarchyRoot);
+	}
+
+	return false;
+}
+
+bool UOwnershipLockingPolicy::IsExplicitlyLocked(const AActor* Actor) const
+{
+	return ActorToLockingState.Contains(Actor);
+}
+
+bool UOwnershipLockingPolicy::IsLockedHierarchyRoot(const AActor* Actor) const
+{
+	return LockedOwnershipRootActorToExplicitlyLockedActors.Contains(Actor);
+}
+
+bool UOwnershipLockingPolicy::AcquireLockFromDelegate(AActor* ActorToLock, const FString& DelegateLockIdentifier)
+{
+	ActorLockToken LockToken = AcquireLock(ActorToLock, DelegateLockIdentifier);
+	if (LockToken == SpatialConstants::INVALID_ACTOR_LOCK_TOKEN)
+	{
+		UE_LOG(LogOwnershipLockingPolicy, Error, TEXT("AcquireLock called from engine delegate returned an invalid token"));
+		return false;
+	}
+
+	check(!DelegateLockingIdentifierToActorLockToken.Contains(DelegateLockIdentifier));
+	DelegateLockingIdentifierToActorLockToken.Add(DelegateLockIdentifier, LockToken);
+	return true;
+}
+
+bool UOwnershipLockingPolicy::ReleaseLockFromDelegate(AActor* ActorToRelease, const FString& DelegateLockIdentifier)
+{
+	if (!DelegateLockingIdentifierToActorLockToken.Contains(DelegateLockIdentifier))
+	{
+		UE_LOG(LogOwnershipLockingPolicy, Error, TEXT("Executed ReleaseLockDelegate for unidentified delegate lock identifier. Token: %s."), *DelegateLockIdentifier);
+		return false;
+	}
+	ActorLockToken LockToken = DelegateLockingIdentifierToActorLockToken.FindAndRemoveChecked(DelegateLockIdentifier);
+	bool bReleaseSucceeded = ReleaseLock(LockToken);
+	return bReleaseSucceeded;
+}
+
+void UOwnershipLockingPolicy::OnOwnerUpdated(const AActor* Actor, const AActor* OldOwner)
+{
+	check(Actor != nullptr);
+
+	// If an explicitly locked Actor is changing owner.
+	if (IsExplicitlyLocked(Actor))
+	{
+		RecalculateLockedActorOwnershipHierarchyInformation(Actor);
+	}
+
+	// If a locked hierarchy root is changing owner.
+	if (IsLockedHierarchyRoot(Actor))
+	{
+		RecalculateAllExplicitlyLockedActorsInThisHierarchy(Actor);
+	}
+	// If an Actor in a locked hierarchy is changing owner (i.e. either the old owner or
+	// the root hierarchy of the old owner is the root of a locked hierarchy), we need to
+	// recalculate ownership hierarchies of all explicitly locked Actors in that hierarchy.
+	else if (OldOwner != nullptr)
+	{
+		const AActor* OldHierarchyRoot = OldOwner->GetOwner() != nullptr ? SpatialGDK::GetHierarchyRoot(OldOwner) : OldOwner;
+		if (IsLockedHierarchyRoot(OldHierarchyRoot))
+		{
+			RecalculateAllExplicitlyLockedActorsInThisHierarchy(OldHierarchyRoot);
+		}
+	}
+ }
+
+void UOwnershipLockingPolicy::OnExplicitlyLockedActorDeleted(AActor* DestroyedActor)
```

Comment:
Do we need to track the root being destroyed also?

Reply:
Oh, you do

Reply:
We do

Reply:
`OnHierarchyRootActorDeleted`